readme.txt
/**
 * Pickles Crawler (PxPlugin)
 ************************************** */

Pickles Crawler は、ウェブサイトを巡回し、
HTMLやリソースファイルを収集するクロールツールです。
Pickles Framework 1.x 系のプラグインとして動作します。

Pickles Framework については、
下記のウェブサイトを参照してください。
http://pickles.pxt.jp/

■インストール方法

1. Pickles Framework をセットアップする。
2. ディレクトリ PicklesCrawler を、
   Pickles Framework の plugins ディレクトリにアップロードする。

■使い方
1. ブラウザで、PxCommand "?PX=plugins.PicklesCrawler" にアクセスする。
2. プロジェクトを作成する。
3. 作成したプロジェクトに、プログラムを作成する。
4. プログラムの「実行」を押し、クロールを開始する。



------
(C)Tomoya Koyanagi.
http://www.pxt.jp/



_docs/changelog.txt
/**
 * PicklesCrawler plugin 更新履歴
 */

[-- PicklesCrawler 1.0.0b4 (2013/**/**) --]

- 「URL変換時に省略するファイル名」の設定を正しくロードできない不具合を修正。


[-- PicklesCrawler 1.0.0b3 (2013/9/11) --]

- スタートページ一覧を大量に投入した場合に、初期設定時点でタイムアウトが発生し得る不具合を修正。
- UIを一部改善。
- PicklesCrawlerディレクトリをplugins下に移動した。


[-- PicklesCrawler 1.0.0b2 (2013/6/8) --]

- PxFW API仕様に則り、プラグインのバージョン番号を返すようにした。
- PxFW 1.0.0b6 での変更に対応した。


[-- PicklesCrawler 1.0.0b1 (2013/5/3) --]

- Pickles Framework 1.x 系に移植。


[-- PicklesCrawler 0.4.3 (2011/08/28) --]
・PicklesFrameworkの拡張HTTPレスポンスヘッダー X-PXFW-RELATEDLINK に対応した。
・プロジェクトにHTML内のJavaScriptを解析するフラグを追加。デフォルトでは解析しないようになった。
・プロジェクトにURL変換時に省略するファイル名を追加。

[-- PicklesCrawler 0.4.2 (2011/04/17) --]
・収集したコンテンツをダウンロードする際のダウンロードファイル名の命名規則を変更。プログラムIDの前に日付を付けるようにした。
・ダイジェスト認証方式に対応した。
・サイト外指定されたパスをURLに変換するオプションを追加。
・クロール時の設定内容を、ログとして残すようになった。

[-- PicklesCrawler 0.4.1 (2010/11/22) --]
・文字コード変換機能で、HTML5の文字コード指定書式に対応した。
・管理画面の表示を、ほんの少し親切にした。

[-- PicklesCrawler 0.4.0 (2010/08/25) --]
・「//」で始まるリンクに対応した。
・「data:」で始まるリンクに対応した。
・プロジェクトの設定をエクスポートする機能を追加。
・text/javascript, application/javascript をパースしないようにした。

[-- PicklesCrawler 0.3.9 (2010/08/11) --]
・収集したファイルを削除する際に、削除するファイル数が多いとタイムアウトが発生して異常終了する問題を修正。

[-- PicklesCrawler 0.3.8 (2010/05/14) --]
・属性値の解析ロジックで、HTMLエンティティを正しくデコードせずに取り扱っていた不具合を修正。

[-- PicklesCrawler 0.3.7 (2010/04/03) --]
・パスの前後に空白文字が入っていてもURLを解析できるようになった。
・イベントハンドラ系の属性が2つ以上指定されていると解析できない不具合を修正。他、属性解析のロジックを改良。
・プロジェクト情報の「対象外URLリスト」と「追加スタートページURLリスト」、プログラム情報の「対象範囲とするURLリスト」と「ダウンロードしないURLリスト」で、=記号(イコール記号)を含むURLを登録できるようになった。
・httpとhttpsの以外のスキーマへのリンクを相対パスとして辿ってしまう問題を修正。
・結果レポートに Last-Modified(更新日) を追加した。
・<object>タグのdata属性に対応した。
・「対象外URLリスト」、「対象範囲とするURL」、「ダウンロード対象外のURL」の評価で、大文字小文字を区別しなくなった。
・その他幾つかの細かい修正。

[-- PicklesCrawler 0.3.6 (2010/01/14) --]
・リンク先パスの置換時に、GETパラメータが消失してしまう不具合を修正。

[-- PicklesCrawler 0.3.5 (2009/12/11) --]
・相対パスでファイル名を省略した(スラッシュで終わる)リンクを検知したときに、階層が一つズレて認識してしまう不具合を修正。

[-- PicklesCrawler 0.3.4 (2009/09/20) --]
・HTTPアクセスオブジェクトを PxHTTPAccess 1.0.1 に更新。取得したコンテンツをファイルに保存せず直接返却させる場合に、リダイレクトの処理が正しく動作しないバグを修正。

[-- PicklesCrawler 0.3.3 (2009/08/30) --]
・プロジェクトに新しい設定項目「複製先パス」を追加。収集完了後に複製する先を指示する。
・プログラムに新しい設定項目「複製先パス」を追加。プロジェクトの同設定項目を上書きする。
・その他幾つかの細かい修正。

[-- PicklesCrawler 0.3.2 (2009/06/23) --]
・文字コード変換機能で、HTMLのメタタグを読み取れないことがある不具合を修正。

[-- PicklesCrawler 0.3.1 (2009/06/06) --]
・CSS の @import句で、メディアタイプの指定が付いている場合に無視してしまう不具合を修正。

[-- PicklesCrawler 0.3.0 (2009/04/04) --]
・<area>タグのhref属性に対応した。
・全てのタグのbackground属性に対応した。
・収集後の文字コード・改行コード変換機能を追加した。
・収集後の一括置換機能を追加した。
・プログラムに、常に送信するパラメータの設定を追加した。
・program フォルダを prg に、downloaded フォルダを dl に改名。短くした。
・クロール結果一覧に「クロールエラー」列を追加。
・パス名/ファイル名が長いために保存できないファイルがあった場合に、クロールエラーを残して通知するようになった。

[-- PicklesCrawler 0.2.2 (2009/03/12) --]
・ダウンロードファイルを一旦一時ファイルとして保存するようになった。
・Not Found ページ収集設定を追加。HTTPステータスが400～500番台の場合にファイルを保存しないオプション。

[-- PicklesCrawler 0.2.1 (2009/03/07) --]
・ファイルを保存する先に既にディレクトリが存在した場合に、無限ループに陥る不具合を修正。

[-- PicklesCrawler 0.2.0 (2009/03/03) --]
・パス指定変換欄で「絶対パス」を選択した場合に、リライトルールが一部適用されない不具合を修正。

[-- PicklesCrawler 0.1.14 (2008/12/10) --]
・プログラム作成時に、予めデフォルト値を自動的にセットするようにした。
・プログラム作成の入力欄にヒントを追加。
・PHPでZipモジュールが利用できる場合に、収集したデータをZIP形式でダウンロードできるようになった。
・クロール中にエラーが発生した場合に、永遠にリトライし続けることがあった不具合を修正。
・その他、管理画面のレイアウトなどを少し変更した。

[-- PicklesCrawler 0.1.13 --]
・HTMLのコメントアウトを無視していた問題を解決。コメント中の文字列を加工しないようになった。
・CSSのコメントアウトを無視していた問題を解決。コメント中の文字列を加工しないようになった。

[-- PicklesCrawler 0.1.12 --]
・拡張子が .css のファイルで、デリミタが付いていない、かつ、同じCSSファイル内に同一パスが2つ以上ある url() 指定に対し、誤った相対パスに変換されることがある不具合を修正。

[-- PicklesCrawler 0.1.11 --]
・tarコマンドを実行する周辺のエスケープ処理をちゃんとした。

[-- PicklesCrawler 0.1.10 --]
・管理画面で、$theme->act() や $theme->href() などを、HTMLエンティティ変換せずに出力していた問題を修正(セキュリティアップデート)。Pickles Framework 0.3.7 の修正に従った。
・その他いくつかの細かい修正。

[-- PicklesCrawler 0.1.9 --]
・フォームのデータを読み込むんだあと、同じページのフォーム以外のURLに対しても同様にパラメータを付加してしまっていた不具合を修正。
・その他いくつかの細かい修正。

[-- PicklesCrawler 0.1.8 --]
・https で始まるURLに対して、SSL通信を行っていなかった不具合を修正。
・通信エラーをレポートファイルに記録するように修正。

[-- PicklesCrawler 0.1.7 --]
・フォーム解析でchecked指定のないチェックボックス、ラジオボタンを送信してしまうことがある不具合を修正。
・プロジェクトにフォームを送信するかどうかの設定項目を追加した。
・フォームデータの送信時に、送信可否の判定をせずに送信していた不具合を修正。
・リライトルールの必須パラメータの評価が、フォームデータを対象としていなかった不具合を修正。
・その他いくつかの細かい修正。

[-- PicklesCrawler 0.1.6 --]
・フォームを検出した場合に、デフォルトの値で送信する機能を実装。
・未定義のパラメータを送信するかどうか、設定できるようにした。
・その他いくつかの修正。

[-- PicklesCrawler 0.1.5 --]
・ポート番号が含まれるURLの場合に、相対パスがズレるバグを修正。

[-- PicklesCrawler 0.1.4 --]
・送信する設定になっているはずのパラメータを送信していなかった不具合を修正。
・その他いくつかのバグを修正。

[-- PicklesCrawler 0.1.3 --]
・管理画面を一部修正。URLリライトルールの上下移動機能およびURLパラメータ定義の自動ソートを追加した。
・ロックファイルを作成し、2重起動を防止するロジックを追加した。ただし、ロックファイルはプログラム単位で生成するため、別のプログラムであれば同時に実行できる。
・実行中のキャンセル要求を受け入れる仕組みを追加。
・その他いくつかの細かい修正。

[-- PicklesCrawler 0.1.2 --]
・サイト外のパスを評価する際に、ドキュメントルートが評価されていなかった不具合を修正。

[-- PicklesCrawler 0.1.1 --]
・プログラムの設定項目に「対象範囲とするURLリスト」を追加。




PicklesCrawler/programbase/execute.php
<?php

#	Copyright (C)Tomoya Koyanagi.
#	Last Update: 14:26 2011/04/17

#******************************************************************************************************************
#	各プログラムの基底クラス
class pxplugin_PicklesCrawler_programbase_execute{

	var $conf;
	var $errors;
	var $req;
	var $dbh;

	var $pcconf;
	var $proj;
	var $program;

	var $result = array();	//結果メッセージを記憶する配列

	var $debug_mode = false;
		#	開発中のデバッグメッセージを出力するか。

	#--------------------------------------
	#	コンストラクタ
	function pxplugin_PicklesCrawler_programbase_execute( &$conf , &$pcconf , &$proj , &$program , &$errors , &$dbh , &$req ){
		$this->conf = &$conf;
		$this->pcconf = &$pcconf;
		$this->proj = &$proj;
		$this->program = &$program;
		$this->errors = &$errors;
		$this->dbh = &$dbh;
		$this->req = &$req;

		$this->additional_constructor();
	}
	#--------------------------------------
	#	コンストラクタの追加処理
	function additional_constructor(){
		#	必要に応じて拡張してください。
	}





	#########################################################################################################################################################
	#	処理の開始
	function execute( &$httpaccess , $current_url , $saved_file_path , $options = array() ){
		/** -------------------------------------- **
			&$httpaccess
				HTTPアクセスオブジェクト
			$current_url
				ダウンロードした現在のURL。
			$saved_file_path
				ダウンロードして保存された先のファイルのフルパス
			$options
				current_url にアクセスした時のオプション(PicklesCrawler 0.1.7 追加)
		/** -------------------------------------- **/
		return	true;
	}

	#--------------------------------------
	#	ロードするURLを新たに追加する処理
	function add_download_url( $add_url , $option = array() ){
		#	$option['referer']//リファラ値;
		#	$option['method'];//送信メソッド(GET/POST)
		#	$option['post'];//POSTメソッドで送る値($option['method']がPOSTじゃない場合は無視される)
		#	$option['type'];//<a>とか<form>とか<img>とか<style>とか<script>とかの区別をしたい。

		$this->msg( 100 , 'ダウンロードリストに追加。' , 'add_download_url' , $add_url , $option );
		return	true;
	}


	#--------------------------------------
	#	エラーをログに残す
	function error_log( $msg = '' ){
		return	$this->errors->error_log( $msg );
	}

	#--------------------------------------
	#	実行結果を出力する
	function msg( $status_cd , $msg = '' , $program = '' , $parameter = '' , $option = array() ){
		/** --------------------------------------
		【　ステータスコード　】
		※PicklesDYプラグイン の API の <progress /> と同じ。
		※$option は、拡張項目です。
		-------------------------------------- **/

		if( $status_cd == 150 && !$this->debug_mode ){
			#	150番はデバッグメッセージ
			#	debug_mode が無効なら出力しない
			return	true;
		}

		$result = array(
			'status'=>intval( $status_cd ),
			'division'=>'crawler',
			'program'=>$program,
			'parameter'=>$parameter,
			'usermessage'=>$msg,
			'option'=>$option,//拡張項目
		);

		array_push( $this->result , $result );
		return	true;
	}

	#--------------------------------------
	#	実行結果を得る
	function get_result(){
		return	$this->result;
	}



	#========================================================================================================================================================
	#	★URL変換系メソッド★
	#* ------------------------------------------------------------------------------------------------------------------ *
	#	これらのメソッドは、ダウンロードしたHTMLから取得したURLの変換を行います。
	#	1つのURLは、処理の間にいくつかの状態を経ます。
	#
	#		ST1■ダウンロードしたソースに書かれている生の状態
	#			絶対パス/相対パス共にありえます。
	#
	#		ST2■HTTPから始まる絶対パスの状態
	#			必ず絶対パスです。
	#			引き継がないパラメータの設定などを考慮して決められるため、
	#			必ずしもST1と同じURLを指すとは限りません。
	#			この状態のURLが、次のダウンロードページとして追加リクエストされます。
	#			ページ内アンカー(#以降)は除去されます。
	#
	#		ST3■ダウンロード後のHTML上で書き換えられる最終的な相対パス
	#			基本的に相対パス/対象がサイト外の場合は絶対パス。
	#			ページ内アンカー(#以降)は保持されます。
	#
	#* ------------------------------------------------------------------------------------------------------------------ *

	#--------------------------------------
	#	URLをパースして、変換指示を返す
	#	$method,$post_data : 19:52 2008/04/16 追加
	function get_replace_url_info( $current_url , $TARGET_PATH_ORIGINAL , $URL_PROTOCOL , $URL_DOMAIN , $current_virtual_url , $method = 'GET' , $FORM_DATA = null ){
		if( !strlen( $method ) ){
			$method = 'GET';
		}
		$method = strtoupper( $method );

		$TARGET_PATH_ORIGINAL = trim( $TARGET_PATH_ORIGINAL );//PxCrawler 0.4.0 : trim() するようにした。
		$TARGET_PATH = $TARGET_PATH_ORIGINAL;
		if( preg_match( '/^(?:[a-zA-Z]+:)?\/\/.*$/is' , $TARGET_PATH ) && !preg_match( '/^'.preg_quote($URL_PROTOCOL,'/').':\/\/'.preg_quote($URL_DOMAIN,'/').'\//is' , $TARGET_PATH_ORIGINAL ) ){
			#	ドメインが別物だったら、除外
			#	PxCrawler 0.4.0 : 「//」 で始まるリンクに対応
			return	false;
		}
		if( preg_match( '/^data:[a-zA-Z0-9\_\-]+\/[a-zA-Z0-9\_\-]+.*$/is' , $TARGET_PATH ) && !preg_match( '/^'.preg_quote($URL_PROTOCOL,'/').':\/\/'.preg_quote($URL_DOMAIN,'/').'\//is' , $TARGET_PATH_ORIGINAL ) ){
			#	「data:」で始まるリンク(HTML埋め込み画像など)に対応
			#	PxCrawler 0.4.0 : 追加
			return	false;
		}

		$anchor = null;
		#	アンカーを考慮
		if( strpos( $TARGET_PATH , '#' ) ){
			#	ターゲットパスから一旦アンカーを削除して、
			#	でも忘れちゃいけないから $anchor にメモる。
			list( $TARGET_PATH , $anchor ) = explode( '#' , $TARGET_PATH , 2 );
		}

		$params = null;
		if( strpos( $TARGET_PATH , '?' ) ){
			#	パラメータがあったら覚えておく
			#	PicklesCrawler 0.1.4 追記
			list( $TARGET_PATH , $params ) = explode( '?' , $TARGET_PATH , 2 );
		}
		if( strlen( $params ) ){
			#	PicklesCrawler 0.3.6 追記
			#	送信してはいけないパラメータが含まれていた場合、
			#	この時点でそぎ落としてしまう。
			$param_list = explode( '&' , $params );
			$alive_param_memo = array();
			foreach( $param_list as $param_line ){
				list( $key , $val ) = explode( '=' , $param_line );
				if( !$this->proj->is_param_allowed( urldecode( $key ) ) ){
					continue;
				}
				array_push( $alive_param_memo , $param_line );
			}
			$params = implode( '&' , $alive_param_memo );
			#	/ PicklesCrawler 0.3.6 追記
		}

		$full_params = $params; //←URLとFORM_DATAをあわせた全パラメータ(20:03 2008/04/16 追加)
		if( strlen( $FORM_DATA ) ){
			$full_params = implode( '&' , array( $full_params , $FORM_DATA ) );
		}

		$TARGET_PATH = $this->optimize_url( $current_url , $TARGET_PATH );//←ここで $TARGET_PATH はrealpathになってるハズ。
		$TARGET_URL = $URL_PROTOCOL.'://'.$URL_DOMAIN.$TARGET_PATH;//←これは、参照対象のURL。

		$REPLACE_TO = '';
#		if( $this->proj->is_outofsite( $TARGET_URL ) ){
#			#	参照対象がサイト外設定されていたら、
#			#	URLに変換し、ダウンロード対象には含まない。
#			$REPLACE_TO = $TARGET_URL;
##			$REPLACE_TO .= '#'.$anchor;//アンカーを復活させる//PxCrawler 0.3.6 削除 (あとで一括して処理してるから)
#		}else{
			$localpath_save_to = $this->proj->url2localpath( implode( '?' , array( $TARGET_URL , $full_params ) ) , $FORM_DATA );
				#	URLを内部パス(保存先パス)に変換する。
				#	$localpath_save_to には、'/http/www.xxx.jp/aaa/index.html' のような文字列が入ります。
				#	$localpath_save_to は、rewrite_rulesが適用されています。
			$URL_DOMAIN = preg_replace( '/\:/' , '_' , $URL_DOMAIN );
			$REPLACE_TO_ABSOLUTE = preg_replace( '/^'.preg_quote( '/'.$URL_PROTOCOL.'/'.$URL_DOMAIN , '/' ).'/' , '' , $localpath_save_to );
			$REPLACE_TO = $this->convert2relativepath( $current_virtual_url , $REPLACE_TO_ABSOLUTE );
#			$REPLACE_TO .= '#'.$anchor;//アンカーを復活させる//PxCrawler 0.3.6 削除 (あとで一括して処理してるから)
//			unset( $REPLACE_TO_ABSOLUTE );
#		}

		if( $method == 'GET' ){
			$ADD_URL = $this->convert2url( $current_url , $TARGET_PATH , $full_params );
		}else{
			$ADD_URL = $this->convert2url( $current_url , $TARGET_PATH , $params );
		}

		#--------------------------------------
		#	回答を作成
		$RTN = array();
		$RTN['replace_to'] = null;
		$path_conv_method = $this->proj->get_path_conv_method();
		if( $this->proj->is_outofsite( $TARGET_URL ) && $this->proj->get_outofsite2url_flg() ){
			$RTN['replace_to'] = $TARGET_URL;
		}else{
			switch( $path_conv_method ){
				#	パス変換方法設定を考慮して変換パスを指示
				case 'absolute':
					if( strlen( $localpath_save_to ) ){
						#	23:47 2009/03/02 Pickles Crawler 0.2.0
						#	参照先が絶対パスにならない問題を修正。
						$RTN['replace_to'] = $REPLACE_TO_ABSOLUTE;
					}else{
						$RTN['replace_to'] = $TARGET_PATH;
					}
					break;
				case 'url':
					$RTN['replace_to'] = $TARGET_URL;
					break;
				case 'none':
					$RTN['replace_to'] = $TARGET_PATH_ORIGINAL;
					break;
				case 'relative':
				default:
					$RTN['replace_to'] = $REPLACE_TO;
					break;
			}
		}
		if( $path_conv_method != 'none' ){
			//「変換しない」設定の場合、この時点で$RTN['replace_to']は完成されているので添加物を加えない。
			if( !preg_match( '/\/$/si' , $RTN['replace_to'] ) ){
				//PxCrawler 0.4.3 追加：省略するファイル名の削除処理
				foreach( $this->proj->get_omit_filename() as $omit_filename ){
					if( basename( $RTN['replace_to'] ) == $omit_filename ){
						$RTN['replace_to'] = preg_replace( '/'.preg_quote($omit_filename,'/').'$/si' , '' , $RTN['replace_to'] );
						if( !strlen( $RTN['replace_to'] ) ){
							$RTN['replace_to'] = './';
						}
						break;
					}
				}
			}
			if( strlen( $params ) ){ $RTN['replace_to'] .= '?'.$params; };//パラメータを復活させる//PxCrawler 0.3.6 追加
			if( strlen( $anchor ) ){ $RTN['replace_to'] .= '#'.$anchor; };//アンカーを復活させる//PxCrawler 0.3.6 追加
		}
		$RTN['add_url'] = $ADD_URL;
		if( strlen( $FORM_DATA ) ){
			if( $method == 'POST' ){
				$RTN['post_data'] = $FORM_DATA;
			}
		}
		return	$RTN;

	}


	#--------------------------------------
	#	リンク先URLの最適化
	function optimize_url( $current_url , $TARGET_PATH ){
		#	絶対パスと相対パスとを吸収し、
		#	スラッシュから始まる絶対パスに変換する。

		if( preg_match( '/^[a-zA-Z]+:\/\/.*?(\/.*)$/is' , $TARGET_PATH , $preg_result ) ){
			#	リンク先がURLになっていたら。
			return	$preg_result[1];
		}

		$urlinfo = parse_url( trim( $current_url ) );
		$DOMAIN = $urlinfo['host'];
		if( strlen( $urlinfo['port'] ) ){
			$DOMAIN .= ':'.$urlinfo['port'];
		}

		$TARGET_PATH = $this->convert2realpath( $current_url , $TARGET_PATH );
		$TARGET_PATH = $this->proj->optimize_url( $urlinfo['scheme'].'://'.$DOMAIN.''.$TARGET_PATH );

		$urlinfo2 = parse_url( trim( $TARGET_PATH ) );

		$RTN = $urlinfo2['path'];
		return	$RTN;
	}


	#--------------------------------------
	#	現在地点(URL)から、$pathへの相対パスを求める
	function convert2relativepath( $url , $path ){

		#	アンカーを考慮
		if( strpos( $url , '#' ) ){
			list($url,$dmy) = explode('#',$url,2);
			unset($dmy);
		}
		if( strpos( $path , '#' ) ){
			list($path,$anchor) = explode('#',$path,2);
		}

		if( !preg_match( '/^([a-zA-Z]+)\:\/\/(.+?(?:\:[0-9]+)?)(\/.*)/i' , $url , $result ) ){
			#	URLの形式が不正ならスキップ
			return	false;
		}
		$PROTOCOL = strtolower( $result[1] );
		$DOMAIN = strtolower( $result[2] );
		$INNER_PATH = $result[3];

		$path = $this->convert2realpath( $url , $path );
			#	↑リンク先が相対パスかも知れないから、
			#	　一旦絶対パス(スラッシュから始まる)に変換している。

		//	Windows形式→UNIX形式
		$DIR_INNER_PATH = preg_replace( '/\\\\/' , '/' , dirname( $INNER_PATH ) );
		$DIR_path = preg_replace( '/\\\\/' , '/' , dirname( $path ) );

		$debug_loop_limit = 5000;
		$debug_loop_count = 0;
		while( strlen( $DIR_INNER_PATH ) && strlen( $DIR_path ) ){
			$debug_loop_count ++;
			if( $debug_loop_count > $debug_loop_limit ){
				#	無限ループ回避のための安全措置
				$errormsg = '無限ループの危険性を察知しました。';
				$this->error_log( $errormsg );
				$this->msg( 500 , $errormsg );
				break;
			}

			$tmp_inner_lay = explode( '/' , $DIR_INNER_PATH );
			array_shift( $tmp_inner_lay );
			$tmp_inner_path = '/'.array_shift( $tmp_inner_lay );

			$tmp_path_lay = explode( '/' , $DIR_path );
			array_shift( $tmp_path_lay );
			$tmp_path = '/'.array_shift( $tmp_path_lay );

			if( $tmp_inner_path != $tmp_path ){
				break;
			}
			unset( $tmp_inner_path , $tmp_path );
			$DIR_INNER_PATH = '';
			if( count( $tmp_inner_lay ) ){
				$DIR_INNER_PATH = '/'.implode( '/' , $tmp_inner_lay );
			}

			$DIR_path = '';
			if( count( $tmp_path_lay ) ){
				$DIR_path = '/'.implode( '/' , $tmp_path_lay );
			}

			unset( $tmp_inner_lay , $tmp_path_lay );

		}

		$path_layers = explode( '/' , $DIR_INNER_PATH );
		$tmp_path_layers_memo = array();
		foreach( $path_layers as $layer ){
			if( !strlen( $layer ) ){ continue; }
			array_push( $tmp_path_layers_memo , '..' );
		}

		$DIR_path = preg_replace( '/^\/+/' , '' , $DIR_path );
		$lay_str = '';
		if( count( $tmp_path_layers_memo ) ){
			$lay_str = implode( '/' , $tmp_path_layers_memo ).'/';
		}

		$RTN = '';
		$RTN .= $lay_str;
		if( strlen( $DIR_path ) ){
			$RTN .= $DIR_path.'/';
		}
		$RTN .= basename( $path );
		if( strlen($anchor) ){
			$RTN .= '#'.$anchor;
		}
		return	$RTN;
	}

	#--------------------------------------
	#	現在地点(URL)から、相対的に解釈された$pathへの絶対的URLを求める
	function convert2url( $url , $path , $params = null ){
		#	$params は、PicklesCrawler 0.1.4 で追加されました。

		if( !preg_match( '/^([a-zA-Z]+)\:\/\/(.+?(?:\:[0-9]+)?)(\/.*)/i' , $url , $result ) ){
			return	false;
		}

		$urlinfo = parse_url( trim( $url ) );
		$PROTOCOL = strtolower( $urlinfo['scheme'] );
		$DOMAIN = strtolower( $urlinfo['host'] );
		if( strlen( $urlinfo['port'] ) ){
			$DOMAIN .= ':'.strtolower( $urlinfo['port'] );
		}
		$INNER_PATH = $urlinfo['path'];

		$path = $this->convert2realpath( $url , $path );

		if( !preg_match( '/^\//' , $path ) ){
			return	false;
		}

		$RTN = $PROTOCOL.'://'.$DOMAIN.$path;
		if( strlen( $params ) ){
			#	PicklesCrawler 0.1.4 追記
			$param_list = explode( '&' , $params );
			$alive_param_memo = array();
			foreach( $param_list as $param_line ){
				list( $key , $val ) = explode( '=' , $param_line );
				if( !$this->proj->is_param_allowed( urldecode( $key ) ) ){
					continue;
				}
				array_push( $alive_param_memo , $param_line );
			}
			$RTN .= '?'.implode( '&' , $alive_param_memo );
		}
		return	$RTN;
	}

	#--------------------------------------
	#	相対パスを絶対パスに書き換える
	function convert2realpath( $url , $relativepath ){
		if( preg_match( '/^\//' , $relativepath ) ){
			#	スラッシュから始まっていたら、
			#	変換の必要はない。
			return	$relativepath;
		}

		if( !preg_match( '/^([a-zA-Z]+)\:\/\/(.+?(?:\:[0-9]+)?)(\/.*)/i' , $url ) ){
			return	false;
		}

		$urlinfo = parse_url( trim( $url ) );
		$PROTOCOL = strtolower( $urlinfo['scheme'] );
		$DOMAIN = strtolower( $urlinfo['host'] );
		if( strlen( $urlinfo['port'] ) ){
			$DOMAIN .= ':'.strtolower( $urlinfo['port'] );
		}
		$INNER_PATH = $urlinfo['path'];

		$RTN = dirname( $INNER_PATH );
		if( $RTN == '/' || $RTN == '\\' ){
			$RTN = '';
		}

		$path_layers = explode( '/' , $relativepath );
		foreach( $path_layers as $layer_name ){

#			if( !strlen( $layer_name ) ){ continue; }
#				↑PxCrawler 0.3.5 : 相対パスでファイル名を省略した場合に、階層がズレるバグの原因だったため削除。

			if( $layer_name == '..' ){
				$RTN = dirname( $RTN );
				if( $RTN == '/' ){
					$RTN = '';
				}
			}elseif( $layer_name == '.' ){
				$RTN = $RTN;
			}else{
				$RTN .= '/'.$layer_name;
			}

		}

		return	$RTN;
	}



	#========================================================================================================================================================
	#	★base_resources_htmlparser からの移植

	#----------------------------------------------------------------------------
	#	HTML属性の解析
	function html_attribute_parse( $strings ){
		preg_match_all( $this->get_pattern_attribute() , $strings , $results );
		for( $i = 0; !is_null($results[0][$i]); $i++ ){
			if( !strlen($results[3][$i]) ){
				$results[4][$i] = null;
			}
			if( $results[2][$i] ){
				$RTN[strtolower( $results[1][$i] )] = $this->html2text($results[2][$i]);
			}else{
				$RTN[strtolower( $results[1][$i] )] = $this->html2text($results[4][$i]);
			}
		}
		return	$RTN;
	}

	#--------------------------------------
	#	タグの属性情報を検出するPREGパターンを生成して返す。
	function get_pattern_attribute(){
		#	属性の種類
		$rnsp = '(?:\r\n|\r|\n| |\t)';
		$prop = '[a-z0-9A-Z_-]+';
		$typeA = '([\'"]?)(.*?)\3';	#	ダブルクオートあり
		$typeB = '[^"\' ]+';					#	ダブルクオートなし

		#	属性指定の式
		$prop_exists = '/'.$rnsp.'*('.$prop.')(?:\=(?:('.$typeB.')|'.$typeA.'))?'.$rnsp.'*/is';

		return	$prop_exists;
	}

	#----------------------------------------------------------------------------
	#	受け取ったHTMLをテキスト形式に変換する
	#	(クラス base_static_text からのコピー)
	function html2text(){
		//	htmlspecialchars_decode()というのもあるが、
		//	PHP5以降からなので、とりあえず使ってない。
		list($TEXT) = func_get_args();
		$TEXT = preg_replace( '/<br(?: \/)?>/' , "\n" , $TEXT );
		$TEXT = preg_replace( '/&lt;/' , '<' , $TEXT );
		$TEXT = preg_replace( '/&gt;/' , '>' , $TEXT );
		$TEXT = preg_replace( '/&quot;/' , '"' , $TEXT );
		$TEXT = preg_replace( '/&amp;/' , '&' , $TEXT );
		return	$TEXT;
	}

}

?>